### PR TITLE
DAMD-57: staging multiple datamodel changes

### PR DESCRIPTION
**Pull request by hassanxp**

### CODE REVIEW

File: python/pfs/datamodel/utils.py [hunk 1]
```diff
@@ -33,15 +29,32 @@ def calculate_pfsDesignId(fiberIds, ras, decs):
...
-        m.update("%d %.0f %.0f" % (fiberId, ra, dec))
 
-    # convert to int and truncate to 8 hexadecimal digits
-    sha1 = int(m.hexdigest(), 16) & 0xffffffff
+def createHash(*args):
+    """Create a hash from the input strings truncated to 63 bits.
+    
+    Parameters
+    ----------
+    args : `str`
```
Comment (PaulPrice) [inline python/pfs/datamodel/utils.py]: `*args`: https://developer.lsst.io/python/numpydoc.html#additional-positional-and-keyword-arguments

File: python/pfs/datamodel/utils.py [hunk 2]
```diff
@@ -33,15 +29,32 @@ def calculate_pfsDesignId(fiberIds, ras, decs):
...
+        input string values used to generate the hash.
 
-    return sha1
+    Returns
+    -------
+    truncatedHash: `int`
+        truncated hash value
+    """
+
```
Comment (PaulPrice) [inline python/pfs/datamodel/utils.py]: No extra line here.

File: python/pfs/datamodel/utils.py [hunk 3]
```diff
@@ -33,15 +29,32 @@ def calculate_pfsDesignId(fiberIds, ras, decs):
...
+    m = hashlib.sha1()
+    for l in list(args):
+        m.update(str(l).encode())
+
+    truncatedHash = int(m.hexdigest(), 16) & 0x7fffffffffffffff
```
Comment (PaulPrice) [inline python/pfs/datamodel/utils.py]: Suggest simply `return`ing this without putting it into a variable and wasting lines.

File: python/pfs/datamodel/utils.py [hunk 4]
```diff
@@ -33,15 +29,32 @@ def calculate_pfsDesignId(fiberIds, ras, decs):
     if (ras == 0.0).all() and (decs == 0.0).all(): # don't check fiberIds as this may be lab data
         return 0x0
 
-    m = hashlib.sha1()
+    sha1 = createHash(["%d %.0f %.0f" % (fiberId, ra, dec) for fiberId, ra, dec in zip(fiberIds, ras, decs)])
```
Comment (PaulPrice) [inline python/pfs/datamodel/utils.py]: Just `return` it.

File: python/pfs/datamodel/utils.py [hunk 5]
```diff
@@ -4,24 +4,20 @@
...
     m = hashlib.sha1()
 
-    for visit in visits:
-        m.update("%d" % (visit))
-
-    # convert to int and truncate to 8 hexadecimal digits
-    sha1 = int(m.hexdigest(), 16) & 0xffffffff
-
+    sha1 = createHash(["%d" % (visit) for visit in visits])
     return sha1
```
Comment (PaulPrice) [inline python/pfs/datamodel/utils.py]: Just `return` it immediately.

File: python/pfs/datamodel/target.py [hunk 6]
```diff
@@ -136,16 +136,11 @@ def calculateExpHash(self, keys=("expId",)):
...
-        hasher = hashlib.sha1()
-        for ident in self.identity:
-            for kk in keys:
-                hasher.update(str(ident[kk]).encode())
-
-        # Convert to int and truncate to 8 hexadecimal digits
-        return int(hasher.hexdigest(), 16) & 0xffffffff
+        hash = createHash([str(indent[kk]).encode() for indent in self.identity for kk in sorted(keys)])
+
+        return hash
```
Comment (PaulPrice) [inline python/pfs/datamodel/target.py]: Just `return` it immediately you calculate it.

File: python/pfs/datamodel/utils.py [hunk 7]
```diff
@@ -33,15 +29,32 @@ def calculate_pfsDesignId(fiberIds, ras, decs):
...
+    
+    Parameters
+    ----------
+    args : `str`
+        input string values used to generate the hash.
 
-    return sha1
+    Returns
+    -------
+    truncatedHash: `int`
```
Comment (PaulPrice) [inline python/pfs/datamodel/utils.py]: Need a space either side of the colon: https://developer.lsst.io/python/numpydoc.html#parameters

File: tests/test_createHash.py [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+from pfs.datamodel.utils import createHash
+
+
+class CreateHashTestCase(unittest.TestCase):
+    """ Tests the pfs.model.utils.createHash()
+    """
+
+    def testList(self):
+
+        # Test a simple set of arguments
```
Comment (PaulPrice) [inline tests/test_createHash.py]: May as well put this comment as the method docstring.

File: tests/test_createHash.py [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+        
+        # Create two lists 
+        listA = ['a', 'b']
+        listB = ['y', 'z']
+
+        # Form a list comprehension from the two lists
+        listC = [a + b for a in listA for b in listB]
+
+        # And test
+        self.assertEqual(createHash(listC), 0x6f0e879d749c2d60)
```
Comment (PaulPrice) [inline tests/test_createHash.py]: All the comments in this method seem unnecessary.

File: python/pfs/datamodel/utils.py [hunk 10]
```diff
@@ -24,9 +24,10 @@ def calculate_pfsDesignId(fiberIds, ras, decs):
         if ras is None and decs is None:
             return 0x0
 
-        raise RuntimeError("Either all or none of fiberId, ra, and dec may be None")
+        raise RuntimeError(
+            "Either all or none of fiberId, ra, and dec may be None")
```
Comment (PaulPrice) [inline python/pfs/datamodel/utils.py]: Why is this changed?

File: python/pfs/datamodel/utils.py [hunk 11]
```diff
@@ -24,9 +24,10 @@ def calculate_pfsDesignId(fiberIds, ras, decs):
...
 
-    if (ras == 0.0).all() and (decs == 0.0).all(): # don't check fiberIds as this may be lab data
+    if (ras == 0.0).all() and (decs == 0.0).all():  # don't check fiberIds as this may be lab data
```
Comment (PaulPrice) [inline python/pfs/datamodel/utils.py]: Should be part of a separate commit, as it's unrelated to the stated goal of this commit.

File: python/pfs/datamodel/utils.py [hunk 12]
```diff
@@ -119,3 +121,19 @@ def astropyHeaderFromDict(metadata):
...
+    """Wraparound number of visits to acceptable range (0-999)
+
+    Parameters
+    ----------
+    nVisit : `int`
+        number of visits
+
+    Returns
+    -------
+    nVisit_wrapped: `int`
```
Comment (PaulPrice) [inline python/pfs/datamodel/utils.py]: Space either side of the colon.

File: python/pfs/datamodel/utils.py [hunk 13]
```diff
@@ -119,3 +121,19 @@ def astropyHeaderFromDict(metadata):
             key = "HIERARCH " + key
         header.append((key, value))
     return header
+
+
+def wraparound_nVisit(nVisit):
```
Comment (PaulPrice) [inline python/pfs/datamodel/utils.py]: Should be `camelCase` rather than `snake_case`: `wraparoundNVisit`. Yeah, doesn't look the best, but that's the standard.

File: python/pfs/datamodel/pfsConfig.py [hunk 14]
```diff
@@ -412,7 +412,7 @@ class PfsConfig(PfsDesign):
     ----------
     pfsDesignId : `int`
         PFI design identifier, specifies the intended top-end configuration.
-    expId : `int`
+    visit0 : `int`
```
Comment (PaulPrice) [inline python/pfs/datamodel/pfsConfig.py]: Commit summary says `visit`, not `visit0`. The commit message should explain the difference.

File: python/pfs/datamodel/target.py [hunk 15]
```diff
@@ -156,7 +156,7 @@ def getIdentity(self, hashKeys=("visit",)):
         identity : `dict`
             Keyword-value pairs identifying these observations.
         """
-        return dict(numExp=wraparound_nVisit(len(self)), expHash=self.calculateExpHash(hashKeys))
+        return dict(numExp=wraparound_nVisit(len(self)), pfsVisitHash=self.calculateExpHash(hashKeys))
```
Comment (PaulPrice) [inline python/pfs/datamodel/target.py]: Not changing the name of `calculateExpHash`?

File: python/pfs/datamodel/utils.py (hunk 13 above)
Comment (hassanxp) [reply python/pfs/datamodel/utils.py]: Ok, good to know `camelCase` is preferred. I used other examples in `datamodel`. Will fix.

File: python/pfs/datamodel/utils.py (hunk 1 above)
Comment (hassanxp) [reply python/pfs/datamodel/utils.py]: Corrected.

File: python/pfs/datamodel/utils.py (hunk 3 above)
Comment (hassanxp) [reply python/pfs/datamodel/utils.py]: Agreed. Corrected.

File: python/pfs/datamodel/utils.py (hunk 5 above)
Comment (hassanxp) [reply python/pfs/datamodel/utils.py]: Corrected.

File: python/pfs/datamodel/utils.py (hunk 11 above)
Comment (hassanxp) [reply python/pfs/datamodel/utils.py]: This was a keyboard-slip. Reverted.

File: python/pfs/datamodel/utils.py (hunk 12 above)
Comment (hassanxp) [reply python/pfs/datamodel/utils.py]: Corrected.

File: python/pfs/datamodel/utils.py (hunk 13 above)
Comment (hassanxp) [reply python/pfs/datamodel/utils.py]: Fixed as suggested.

File: python/pfs/datamodel/pfsConfig.py (hunk 14 above)
Comment (hassanxp) [reply python/pfs/datamodel/pfsConfig.py]: Commit message updated in the case of `PfsConfig`.

File: python/pfs/datamodel/target.py (hunk 15 above)
Comment (hassanxp) [reply python/pfs/datamodel/target.py]: Agreed. Renamed function to `calculateVisitHash`.
